overlays/wrap_memory.py

```diff
@@ -0,0 +1,252 @@
+    def produce_value(key: str, upstream_get: Any, current_env_getter: Any) -> T:
+        "Must be implemented by child environments"
+        raise NotImplementedError()
+
+    def register_dependency(self, key: str, dependency: str) -> None:
+        self.writable_env.dependencies[key] = self.writable_env.dependencies.get(key, set())
+        self.writable_env.dependencies[key].add(dependency)
+
+    def get(self, key: str, dependency: str, use_saved_contents_of_dependents: bool) -> T:
+        self.register_dependency(key, dependency)
+
+        target_cache_table = (self.writable_env.saved_contents_cache_table
+                              if use_saved_contents_of_dependents or module(key) not in self.writable_env.unsaved_modules
+                              else self.writable_env.unsaved_contents_cache_table)
+        # Update the saved_contents_cache_table whether the module is
+        # saved or unsaved.
+        if key not in target_cache_table:
+            target_cache_table[key] = self.produce_value(
+                key,
+                self.upstream_get(use_saved_contents_of_dependents),
+                current_env_getter=target_cache_table.get
+            )
+
+        return target_cache_table[key]
+
+    def update_for_push(self, keys_to_update: Set[str], is_saved_content: bool) -> Set[str]:
+        def update_table(table: Dict[str, T], key: str, use_saved_contents_of_dependents: bool) -> None:
+            table[key] = self.produce_value(
+                key,
+                self.upstream_get(use_saved_contents_of_dependents=use_saved_contents_of_dependents),
+                current_env_getter=table.get
+            )
+
+        downstream_deps = set()
+        for key in keys_to_update:
+            is_unsaved_module = module(key) in self.writable_env.unsaved_modules
+
+            if is_saved_content:
+                # Update saved_contents_cache_table whether the module is saved or unsaved.
+                update_table(self.writable_env.saved_contents_cache_table, key,
+                             use_saved_contents_of_dependents=True)
+
+                # If the module is unsaved, also update
+                # unsaved_contents_cache_table. Newly-saved content needs to
+                # propagate to both saved and unsaved cache tables.
+                if is_unsaved_module:
+                    update_table(self.writable_env.unsaved_contents_cache_table, key,
+                                 use_saved_contents_of_dependents=False)
+            elif is_unsaved_module:
+                update_table(self.writable_env.unsaved_contents_cache_table, key,
+                             use_saved_contents_of_dependents=False)
+
+            downstream_deps |= self.writable_env.dependencies.get(key, set())
+
+        return downstream_deps
+
+    def update(self, module: str, code: str, is_saved_content: bool) -> Set[str]:
+        if is_saved_content:
+            self.writable_env.unsaved_modules.discard(module)
```

Comment:
This bit could be tricker in production since the saved code doesn't necessarily match unsaved editor state.

It's worth noting that we're not clearing out the stale data in the unsaved cache here. I'm not sure whether it can cause bugs to have invalid stale data sitting around if we fail to clear it... I think maybe not, if all of the push-and-pull based logic is perfect.